src/rules/noWarningCommentsRule.ts

```diff
@@ -0,0 +1,175 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as tsutils from "tsutils";
+import * as ts from "typescript";
+import * as Lint from "../index";
+import * as utils from "../utils";
+
+interface IOptionsObject {
+    location?: string;
+    terms?: string[];
+}
+
+interface Options {
+    location?: "start" | "anywhere";
+    terms?: string[];
+}
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-warning-comments",
+        description: "Enforces preventing comments that include any of the terms deemed to be 'warnings'.",
+        rationale: "Helps avoid unfinished or buggy code being checked in to source control.",
+        optionsDescription: Lint.Utils.dedent`
+            Two arguments may be optionally provided:
+
+            * \`"location"\` specifies where to look for 'warning' terms in comments:
+                * 'start' specifies at the start.
+                * 'anywhere' specifies anywhere in the comment.
+            * \`"terms"\` specifies a list of terms (case-insensitive) to be considered 'warning' terms.
+
+            By default, \`"location"\` defaults to \`"start"\` and \`"terms"\` defaults to
+            \`["todo", "fixme", "xxx"]\`.
+            `,
+        options: {
+            type: "object",
+            additionalProperties: false,
+            properties: {
+                terms: {
+                    type: "array",
+                    items: {
+                        type: "string",
+                    },
+                },
+                location: {type: "string"},
+            },
+        },
+        optionExamples: [
+            true,
+            [true, {location: "anywhere"}],
+            [true, {location: "start", terms: ["todo", "fixme", "xxx"]}],
+        ],
+        type: "style",
+        typescriptOnly: false,
+    };
+    /* tslint:enable:object-literal-sort-keys */
+
+    public static COMMENT_ERROR_FACTORY = (term: string, location: "start" | "anywhere"): string =>
+        `Warning term '${term}' found ${location === "start" ? "at start of" : "in"} comment`
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk, parseOptions(this.ruleArguments[0] as IOptionsObject));
+    }
+}
+
+function parseOptions(options: IOptionsObject): Options {
+    if (typeof options === "undefined" || options === null) {
```

Comment:
no need to use `typeof`, please prefer `options == undefined`

Reply:
Why is that?  I usually use this to ensure there's no risk of a redefined `undefined` breaking things.

Reply:
that's not a problem in typescript code. TypeScript warns about declaring a variable named `undefined` or assigning something to `undefined`